CAST_SPELL script commands

kxvv opened the pull request:
-CAST_SPELL
-CAST_SPELL_ON_CREATURE
-CAST_SPELL_AT_LOCATION

AdamPlenty reviewed (commented):
script_cast_spell_on_creature, script_cast_spell_at_location, and script_cast_spell all return a TbBool, but you've specified a TbErrorCode (which is not the same) as a return value. Will this work? Given that a TbBool is the same as an unsigned char (I think), Lb_FAIL might underflow and be interpreted as true.

kxvv commented:
I believe I was using
return magic_use_power_... >= Lb_OK;
at some point and missed to change return type.  Where are those return values used? I see script_... functions being only called in one place and return value is always ignored.

AdamPlenty commented:
> 
> 
> I believe I was using
> return magic_use_power_... >= Lb_OK;
> at some point and missed to change return type. Where are those return values used? I see script_... functions being only called in one place and return value is always ignored.

Lb_OK is 0 and will therefore be interpreted as false. If you mean false, that's what you should put. I'm not sure why these functions are returning a value if it's never used, but you should still make sure it's correct.

kxvv commented:
> > I believe I was using
> > return magic_use_power_... >= Lb_OK;
> > at some point and missed to change return type. Where are those return values used? I see script_... functions being only called in one place and return value is always ignored.
> 
> Lb_OK is 0 and will therefore be interpreted as false. If you mean false, that's what you should put. I'm not sure why these functions are returning a value if it's never used, but you should still make sure it's correct.

Yeah but 0 >= 0 is true so Lb_OK and Lb_SUCCESS would pass while Lb_FAIL wouldn't.
Anyway I'll change the return type, anything else to change?

AdamPlenty commented:
> 
> 
> > > I believe I was using
> > > return magic_use_power_... >= Lb_OK;
> > > at some point and missed to change return type. Where are those return values used? I see script_... functions being only called in one place and return value is always ignored.
> > 
> > 
> > Lb_OK is 0 and will therefore be interpreted as false. If you mean false, that's what you should put. I'm not sure why these functions are returning a value if it's never used, but you should still make sure it's correct.
> 
> Yeah but 0 >= 0 is true so Lb_OK and Lb_SUCCESS would pass while Lb_FAIL wouldn't.
> Anyway I'll change the return type, anything else to change?

Is TbBool is unsigned, Lb_FAIL would probably be interpreted as true.

You might want to have script_cast_spell_at_location check if the coördinates given are valid if the magic_use_power functions don't already.

Loobinex commented:
Thanks KIX, welcome aboard. I'll test your work and look at your code at a later time and will provide feedback. In the meantime, would you be so kind as to explain the script parameters. For example like [this](https://github.com/dkfans/keeperfx/wiki/New-and-Modified-Level-Script-Commands#change_creature_owner) here, so I know exactly how to use them?

SimLV commented on src/lvl_script.c:
```diff
@@ -4119,6 +4214,162 @@ TbBool script_level_up_creature(PlayerNumber plyr_idx, long crmodel, long criter
+TbBool script_cast_spell_on_creature(PlayerNumber plyr_idx, long crmodel, long criteria, long magic_caster_splvl)
+{
+    struct Thing *thing;
+    switch (criteria)
+    {
+    case CSelCrit_Any:
+        thing = get_random_players_creature_of_model(plyr_idx, crmodel);
+        break;
+    case CSelCrit_MostExperienced:
+        thing = find_players_highest_level_creature_of_breed_and_gui_job(crmodel, CrGUIJob_Any, plyr_idx, 0);
+        break;
+    case CSelCrit_MostExpWandering:
+        thing = find_players_highest_level_creature_of_breed_and_gui_job(crmodel, CrGUIJob_Wandering, plyr_idx, 0);
+        break;
+    case CSelCrit_MostExpWorking:
+        thing = find_players_highest_level_creature_of_breed_and_gui_job(crmodel, CrGUIJob_Working, plyr_idx, 0);
+        break;
+    case CSelCrit_MostExpFighting:
+        thing = find_players_highest_level_creature_of_breed_and_gui_job(crmodel, CrGUIJob_Fighting, plyr_idx, 0);
+        break;
```
It would be good if you extract this into a function like
`struct Thing *script_get_creature_by_criteria(PlayerNumber plyr_idx, long criteria);`

SimLV commented on src/lvl_script.c:
```diff
@@ -4119,6 +4214,162 @@ TbBool script_level_up_creature(PlayerNumber plyr_idx, long crmodel, long criter
+    long splevel = (magic_caster_splvl & ((1 << 8) - 1));
+    MapSubtlCoord thingXPos = thing->mappos.x.stl.num;
+    MapSubtlCoord thingYPos = thing->mappos.y.stl.num;
+    switch (powerKind)
+    {
+        case PwrK_DISEASE:
+            return magic_use_power_disease(caster, thing, 0, 0, splevel, PwMod_CastForFree);
+        case PwrK_SPEEDCRTR:
+            return magic_use_power_speed(caster, thing, 0, 0, splevel, PwMod_CastForFree);
+        case PwrK_SLAP:
+            return magic_use_power_slap_thing(caster, thing, PwMod_CastForFree);
+        case PwrK_CHICKEN:
+            return magic_use_power_chicken(caster, thing, 0, 0, splevel, PwMod_CastForFree);
+        case PwrK_CONCEAL:
+            return magic_use_power_conceal(caster, thing, 0, 0, splevel, PwMod_CastForFree);
+        case PwrK_HEALCRTR:
+            return magic_use_power_heal(caster, thing, 0, 0, splevel, PwMod_CastForFree);
+        case PwrK_PROTECT:
+            return magic_use_power_armour(caster, thing, 0, 0, splevel, PwMod_CastForFree);
+        case PwrK_LIGHTNING:
+            return magic_use_power_lightning(caster, thingXPos, thingYPos, splevel, PwMod_CastForFree);
+        case PwrK_CAVEIN:
+            return magic_use_power_cave_in(caster, thingXPos, thingYPos, splevel, PwMod_CastForFree);
+        case PwrK_CALL2ARMS:
+            return magic_use_power_call_to_arms(caster, thingXPos, thingYPos, splevel, PwMod_CastForFree);
+        case PwrK_MKDIGGER:
+            return magic_use_power_imp(caster, thingXPos, thingYPos, PwMod_CastForFree);
+        default:
```
You may use `magic_use_power_on_thing` instead of this big switch

Loobinex replied:
He mentioned it below, but he included spells that are not cast on things. This way the mapmaker can cast for example the sight spell on a specific enemy unit, revealing it.

kxvv commented:
Will do. Didn't notice there's a function for that, I'll also use magic_use_power_on_subtile.

kxvv commented:
Thanks for the tips. Modified the return values. Used <code>magic_use_power_on_level</code> and <code>magic_use_power_on_subtile</code>. I looked at <code>magic_use_power_on_thing</code> but it does not do what I want so I'll stick with the switch. The standalone <code>magic_use_power</code> functions perform the checks I need. Removed some duplicities in criteria based script commands and replaced them with function call. Script commands should work like this:

Script params explanation:
<b>CAST_SPELL_ON_CREATURE([player],[creature],[criterion],[caster_player],[spell],[spell_level])</b>
The first three params as in KILL_CREATURE. It's using the same code
[player] - The player who owns creatures to be affected.
[creature] - Creature model to be searched.
[criterion] - Creature selection criterion.
[caster_player] - The player who casts the spell.
[spell_name] - same as in magic_available command. Cannot use power_hand or power_possess or power_timebomb
[spell_level] - 1 to 9 integer. If spell level is undefined for the spell, it has no effect. Has to be within 1 to 9, otherwise adjusted. (0-8 internally).

<b>CAST_SPELL_AT_LOCATION([caster_player],[subtile_x],[subtile_y],[spell_name],[spell_level])</b>
[subtile_x] - x position of location's subtile
[subtile_y] - y position of location's subtile

<b>CAST_SPELL([caster_player],[spell_name])</b>

Loobinex commented:
The CAST_SPELL_ON_CREATURE does not check if the player is in hand or not. 

I tested it with this script:
```
SET_TIMER(PLAYER0,TIMER1)
IF(PLAYER0,TIMER1 >= 40)
  NEXT_COMMAND_REUSABLE
  CAST_SPELL_ON_CREATURE(PLAYER0,BILE_DEMON,MOST_EXPERIENCED,PLAYER1,POWER_LIGHTNING,8)
  NEXT_COMMAND_REUSABLE
  SET_TIMER(PLAYER0,TIMER1)
ENDIF
```
If you pick up the strongest bile demon, it will continue to cast at the location you picked it up from.

Loobinex commented on src/lvl_script.c:
```diff
@@ -2506,6 +2510,100 @@ void command_level_up_creature(long plr_range_id, const char *crtr_name, const c
+    return;
+  }
+  PowerKind pwr = mag_id;
+  if((PlayerNumber) caster_plyr_idx > PLAYER3)
```
What happens when you allow white to cast these spells?

kxvv replied:
White and neutral lightning/CTA crash the game. So those vals + ALL_PLAYERS are forbidden

Loobinex commented on src/lvl_script.c:
```diff
@@ -4119,6 +4117,90 @@ TbBool script_level_up_creature(PlayerNumber plyr_idx, long crmodel, long criter
+        return magic_use_power_lightning(caster, stl_x, stl_y, splevel, PwMod_CastForFree);
+      case PwrK_CAVEIN:
+        return magic_use_power_cave_in(caster, stl_x, stl_y, splevel, PwMod_CastForFree);
+      case PwrK_MKDIGGER:
```
Funny, these spells aren't meant to be cast at units of course, but it works, so why not I guess. Needs proper documentation.

kxvv replied:
Yeah magic_use_power_on_thing does not list them. But I thought it would be great if player could do this via script.

Loobinex commented:
> Script commands should work like this:
> CAST_SPELL_ON_CREATURE([player],[creature],[criterion],[caster_player],[spell_name],[spell_level])
> CAST_SPELL_AT_LOCATION([caster_player],[subtile_x],[subtile_y],[spell_name],[spell_level])
> CAST_SPELL([caster_player],[spell_name])

I see for each of those, you end up adding the 'cast for free' flag. I can imagine that sometimes a mapmaker would want to charge the player for the privilege of casting a spell through the script, and although you can script this out, wouldn't it be smart to change the script commands to this:

```
CAST_SPELL_ON_CREATURE([player],[creature],[criterion],[caster_player],[spell_name],[spell_level],[free])
CAST_SPELL_AT_LOCATION([caster_player],[subtile_x],[subtile_y],[spell_name],[spell_level],[free])
CAST_SPELL([caster_player],[spell_name],[free])
```

Where if you set the last property to 0, the player is charged for the spell.

SimLV commented:
>  a mapmaker would want to charge
Lets script cast `lightning` on your own heart each time you have more than 1000 gold for your own funds.
_Avarice is evil_

kxvv commented:
Yeah, I thought about it but decided not to. I can add [free] param I guess. 
Hmm ok I'll make it the spell will not be cast on held units.

Loobinex commented:
Well, please share why you decided against it.

kxvv commented:
I imagined some player getting frustrated and wondering where does his money. And it's scripted lvl 9 lightning being cast periodically in non relevant area. On his account.
I'm actually starting to like the idea. There's no other way to drain player's money via script, right?

Loobinex commented:
Sure there is, you can add a negative amount of gold to a player.
I was thinking more along the lines of helping a computer player out through the script, but not for free. To get a scripted, but not cheating, enemy CP.

But sure, messing with the player would work.

kxvv commented:
I believe adding negative amount of money to ADD_GOLD_TO_PLAYER only substracts the 'offmap' money - made by selling (also start money). It doesn't touch the gold you have in gold pots in treasure room.

Loobinex commented:
ok, so yes, this would be the only way to drain treasure rooms.

Loobinex commented:
O, missed it. Sure, will look at it tonight.
